### PR TITLE
Percentile speed up and standard error update

### LINKED ISSUE #1775: [TASK] EconomicRatio consistent with BasicStatistics
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
Value at risk and expected shortfall are calculated similar to percentile in BasicStatistics. #1771 and #1772 updated how percentile metadata is stored (#1771) and calculated when pivot parameters are present (#1772).

**Describe the solution you'd like**
Ensure that value at risk and expected shortfall behave the same as percentile.

**Describe alternatives you've considered**
none

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #1779: [TASK] Percentile and percentile standard error speed up/default behavior
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
The current implementation of percentile and percentile standard error uses three for loops (target, pivot parameter, percent) when calculating the metric for HistorySet. The addition of percentile standard error results in an increased computation time. This is seen in the increase in time to run the plugin test: SR2ML/tests/sept_milestone (from ~130s before implementing standard error to ~480s after and reduced to ~335s when switching loop order in PR #1776).

**Describe the solution you'd like**
Vectorize the innermost loop to speed up the calculation of both percentile and percentile standard error by reducing the number of for loops to two from three. If this is not considered fast enough, implement a default standard error calculation using the normal distribution assumption while allowing the user to select the more accurate KDE implementation.

**Describe alternatives you've considered**
Replace KDE standard error implementation with less accurate but faster equation assuming normal distribution.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by dgarrett622**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1779 
Closes #1775 

##### What are the significant changes in functionality due to this change request?

- Use xarray.DataSet.quantile() method whenever possible (median and percentile when no probability weights given and when probability weights are all equal)
- Speed up calculation of weighted percentiles in _computeWeightedPercentile (xarray native methods do not support weighted quantiles/percentiles)
- Add user parameter 'interpolation' to median, percentile, and value at risk
- Change percentile standard error to standard normal formulation because KDE implementation increases computation time too much when requesting multiple percentiles on multiple variables
- Makes EconomicRatio compatible with BasicStatistics
- Change value at risk standard error to standard normal formulation because KDE implementation increases computation time too much

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by dgarrett622** [comment]
@wangcj05 The computation should be significantly faster now. Is this fast enough, or do we need to implement the standard normal method?

Before merging, I would like to include some notes in the documentation about the sample size needed for percentile and its standard error calculation from the paper.

**Comment by wangcj05** [comment]
@dgarrett622 Thanks for updates. I have checked the test, it reduced the time to ~170s now. Original test took about ~130s. I think the changes would be good for now. But I would like to keep the issue open. The only thing I'm worried about is the nonlinear increase in time. We may need to add a benchmark test in the future to prevent this behavior.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (wangcj05) [inline framework/Models/PostProcessors/BasicStatistics.py:1017]: In our code, we are using equivalentSamples, which are precomputed. We may merge equivalentSamples and self.sampleSize in the future.  Could you update this line? the following is an example:
```
    metric = 'sigma'
    if len(needed[metric]['targets'])>0:
      self.raiseADebug('Starting calculate standard error on "'+metric+'"...')
      varList = list(needed[metric]['targets'])
      if self.pbPresent:
        en = calculations['equivalentSamples'][varList]
        factor = 2.0 * (en - 1.0)
        factor = self._computePower(0.5,factor)
      else:
        factor = np.sqrt(2.0 * (float(self.sampleSize) - 1.0))
      calculations[metric+'_ste'] = calculations['sigma'][varList] / factor
```

**Review by wangcj05 — CHANGES_REQUESTED** [review]
I have some comments for you consideration.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py:615]: "_computeSingleWeightedPercentile" is there to avoid copy/pasting the same code in the if..else.. statement. If I had written this originally, I would have used numpy.interp which would take float, list, or array and would have avoided writing "_computeSingleWeightedPercentile" and the if..else.. statement. The only reason I did not implement this is that it changes the way the percentiles are calculated (linear interpolation instead of the midpoint between the two nearest data points) and would require updating many gold files. I think numpy.interp would speed the calculation because the for loop would be eliminated. 

I can change the "_computeWeightedPercentile" function to accept a list of percents instead of the accepting float or list. If you think it would be worth it to do the numpy.interp method, I could try that as well.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py:994]: Yes, the calculation of quantiles originally required an argsort every time a new percentile was calculated. I moved the argsort outside of this so that the data is sorted once and then multiple quantiles are calculated from that sorted data. This scaled roughly linearly with the number of percents requested.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py:1026]: We cannot avoid this when there are probability weights present. This is because xarray does not currently support weighted quantiles. Without going too far down the rabbit hole, here are some of the issues being worked to enable that functionality:

[https://github.com/pydata/xarray/issues/1371](https://github.com/pydata/xarray/issues/1371)
[https://github.com/numpy/numpy/issues/8935](https://github.com/numpy/numpy/issues/8935)
[https://github.com/numpy/numpy/pull/9211](https://github.com/numpy/numpy/pull/9211)

This is why we have to handle weighted percentiles on our own, which requires looping over targets and time for percentile and percentile standard error. I have made it so that we no longer have to loop over a third parameter (percent) as well.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py:1017]: I did not realize this was available. I'll make the change.

**Comment by dgarrett622** [comment]
I discovered some additional enhancements for speed when using _computeWeightedPercentile. It will involve using xarray.DataSet methods directly when probability weights are provided but are all equal. This will likely require new gold files for some tests. Since some gold files will already need to be updated, I'll update _computeWeightedPercentile with numpy.interp to (hopefully) provide a faster computation.

The KDE implementation is more accurate, however, it takes far too long to compute for even moderately sized samples. I'll implement the standard normal formulation of the standard error for speed.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py]: "_computeWeightedPercentile" has been changed to accept a list of percents. With the new refactoring, "_computeSingleWeightedPercentile" is no longer necessary.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py]: The calculation of quantiles was slowed down by using "_computeWeightedPercentile" instead of using native xarray.DataSet.quantile() methods. I changed this so that if there are no probability weights present or if all the probability weights are equal native xarray.DataSet.quantile() is used. This speeds up the calculation of percentiles significantly. At this time, xarray does not have support for weighted quantiles, so if probability weights are present and they are not all the same, we have to use "_computeWeightedPercentile." The new formulation of "_computeWeightedPercentile" is faster than before, but is still slower than native xarray.DataSet.quantile().

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py]: By switching to the standard normal formulation of the standard error, we no longer have to loop through targets, time, or percent. This speeds up the calculation significantly. I left the KDE implementation commented out in the code in case we want to revisit it.

**Comment by dgarrett622** [comment]
I made a number of changes that significantly speed up the percentile and percentile standard error calculation. Whenever possible native xarray.DataSet.quantile() is used. This yields a slightly different value than before for many cases. I had to update the gold files for many tests. I suspect this will impact a number of plugin tests that use percentile and median.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (wangcj05) [inline framework/Models/PostProcessors/BasicStatistics.py:609]: I assume the difference between the percentile calculation is: originally we compute the average between two data points, the new modification changes it to use the interpolation which is also based on the values of weights CDF. I do have a question here. For example,  we have some discussion on how to compute the weighted median in the past. Some suggestion from (https://en.wikipedia.org/wiki/Weighted_median) using the mean of lower and upper weighted median. Others argue to use the interpolated value to represent the median, as you did in your modification. I'm not the expert in the definition about the weighted median, or weighted percentile. You may take a look at the literatures, and tell us what's your thought, and we can make a decision on Monday meeting. @dgarrett622

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@dgarrett622 Some comments for your consideration. Let's place this PR on hold until next Monday.

File: doc/user_manual/postprocessor.tex
Comment (dgarrett622) [reply doc/user_manual/postprocessor.tex:169]: The changes in the code are a simplification of the asymptotic variance method. We are using the asymptotic variance method with the assumption of a normal distribution. The benefit is that this assumption gives an explicit equation that we can use. I don't think we need to make changes in the documentation, but if you would like me to remove the asymptotic variance description, I can.

File: framework/Models/PostProcessors/BasicStatistics.py
Comment (dgarrett622) [reply framework/Models/PostProcessors/BasicStatistics.py:609]: Sure, I'll discuss my reasoning on Monday.

**Comment by dgarrett622** [comment]
Adding the user specified 'interpolation' parameter to median and percentile affected _computeWeightedPercentile which is used in EconomicRatio. I rolled the fixes for #1775 into this PR since EconomicRatio required updates with this new 'interpolation' parameter.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
Several of comments for your consideration

**Comment by wangcj05** [comment]
@dgarrett622 One more comment, please update some tests to reflect the option of "interpolation=midpoint". In this case, you will test both situations.  Also an email to the user group need to be drafted to reflect the changes.

**Comment by dgarrett622** [comment]
Email draft for users:

Dear users,

PR #1780 makes a number of changes to the BasicStatistics and EconomicRatio PostProcessors that change some of the calculated statistics. This PR introduces the following changes:

- median and percentile in BasicStatistics now have an additional parameter, "interpolation," that can take the values "linear" or "midpoint." This parameter influences how RAVEN interpolates between data points to calculate the median or percentile. The default is "linear." To get the previous behavior, use "midpoint."
- valueAtRisk and expectedShortfall in EconomicRatio have the same "interpolation" parameter as median and percentile.
- Due to the long computational time to calculate the standard error for percentile and valueAtRisk, the standard error method has been changed. The new method is much faster, however, gives slightly different values.
- xarray has been used wherever possible to speed up the computations. This may result in slightly different values for median, percentile, and valueAtRisk.

File: tests/framework/PostProcessors/BasicStatistics/gold/basicStatsMonteCarloAnalytic/analyticalTest_basicStatPP_dump.csv
Comment (wangcj05) [inline tests/framework/PostProcessors/BasicStatistics/gold/basicStatsMonteCarloAnalytic/analyticalTest_basicStatPP_dump.csv:2]: @dgarrett622 I have a question here. Since you have switched the calculation to "midpoint", why the calculation is different from the current devel version? I mean why do we need to regold it also?

Comment (dgarrett622) [reply tests/framework/PostProcessors/BasicStatistics/gold/basicStatsMonteCarloAnalytic/analyticalTest_basicStatPP_dump.csv:2]: @wangcj05 I looked a little more into this discrepancy. When using _computeWeightedPercentile for a variable with N values, the smallest value is appended with weight zero. Adding in this additional value appears to cause the discrepancy with xarray. _computeWeightedPercentile works with N+1 values where xarray works with N values. For many requested percentiles both methods return the same value. However because the two methods use arrays of different length, they may return different values. I get the same value for the 5th percentile using each method and differing values for the 95th percentile when using the data from this test. As another example, when requesting a very small percentile (1e-9) _computeWeightedPercentile will return the smallest value where xarray returns the average of the two smallest values.

The differences in this particular example are small, on the order of 0.06% difference between the values. As written, the majority of cases will be calculated using xarray. The only time _computeWeightedPercentile would be used is if the probability weights are not all the same (xarray does not currently support weighted quantile calculation, I think it is because numpy does not have it implemented). I still think using the standard library xarray wherever possible is the best way forward.

Comment (wangcj05) [reply tests/framework/PostProcessors/BasicStatistics/gold/basicStatsMonteCarloAnalytic/analyticalTest_basicStatPP_dump.csv:2]: Sounds good to me. Thanks for checking it.

**Review by wangcj05 — APPROVED** [review]
PR is good.

**Comment by wangcj05** [comment]
@dgarrett622 I have merged your PR, could you send the drafted email to raven user group?

**Pull request merged**
